### PR TITLE
Added spark profiles to build 2.2/3/4 and 3.0

### PR DESCRIPTION
**Pull request by radford1**
*Issue #, if available:*
#283
*Description of changes:*
Added spark profiles piggy backing off of @alexdemidov work for implementing with scala 2.12

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by radford1** [comment]
Spark 3.0 is passing now

File: pom.xml [hunk 1]
```diff
@@ -107,33 +115,33 @@
             <artifactId>scala-compiler</artifactId>
             <version>${scala.version}</version>
             <scope>test</scope>
-         </dependency>
+        </dependency>
 
         <dependency>
```
Comment (sscdotopen) [inline pom.xml:120]: Why do we need to change the versions of these dependencies? Is it related to the changes for Spark/Scala?

File: pom.xml [hunk 2]
```diff
@@ -377,6 +385,38 @@
...
+        <profile>
+            <id>scala-2.12</id>
+            <properties>
+                <scala.major.version>2.12</scala.major.version>
+                <scala.version>${scala.major.version}.10</scala.version>
+                <artifact.scala.version>_${scala.major.version}</artifact.scala.version>
+            </properties>
+        </profile>
+        <profile>
+            <id>spark-2.2</id>
```
Comment (sscdotopen) [inline pom.xml:397]: Why do we need to include all these Spark 2.x versions? Spark 2.x versiosn should be compatible, so its fine to compile with one 2.x version and run on a cluster with another version. That should work in the current code base.

File: src/main/scala/com/amazon/deequ/analyzers/applicability/Applicability.scala [hunk 3]
```diff
@@ -140,7 +141,14 @@ private[deequ] object Applicability {
     if (shouldBeNull(nullable)) {
       null
     } else {
-      new Timestamp(Random.nextLong())
+      val Low = 100
```
Comment (sscdotopen) [inline src/main/scala/com/amazon/deequ/analyzers/applicability/Applicability.scala:144]: How is this change related to the Spark/Scala update?

File: src/main/scala/com/amazon/deequ/constraints/AnalysisBasedConstraint.scala [hunk 4]
```diff
@@ -1,18 +1,18 @@
...
- *
- *     http://aws.amazon.com/apache2.0/
- *
- * or in the "license" file accompanying this file. This file is distributed on
- * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
- * express or implied. See the License for the specific language governing
- * permissions and limitations under the License.
- *
- */
+  * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (sscdotopen) [inline src/main/scala/com/amazon/deequ/constraints/AnalysisBasedConstraint.scala:2]: How is this change related to the Spark/Scala update?

File: src/main/scala/com/amazon/deequ/constraints/AnalysisBasedConstraint.scala [hunk 5]
```diff
@@ -40,10 +40,10 @@ import scala.util.{Failure, Success, Try}
   *
   */
 private[deequ] case class AnalysisBasedConstraint[S <: State[S], M, V](
-    analyzer: Analyzer[S, Metric[M]],
-    private[deequ] val assertion: V => Boolean,
-    private[deequ] val valuePicker: Option[M => V] = None,
-    private[deequ] val hint: Option[String] = None)
+                                                                        analyzer: Analyzer[S, Metric[M]],
```
Comment (sscdotopen) [inline src/main/scala/com/amazon/deequ/constraints/AnalysisBasedConstraint.scala:43]: How is this change related to the Spark/Scala update?

**Review by sscdotopen — CHANGES_REQUESTED** [review]
Hi,

I started to have a look at the changes, and I am a bit confused, see my comments. Why do we need to update the testing dependencies as well?

File: pom.xml (hunk 1 above)
Comment (radford1) [reply pom.xml:120]: Making it work with scala 2.12

File: pom.xml (hunk 2 above)
Comment (radford1) [reply pom.xml:397]: Wanted to include the other versions in the profile for testing each version. It was mentioned in PR #253 that some more testing should be done. This is building off of that pr and including spark 3 compatibility. If it's overkill we can remove the profiles and keep only 2.4 and 3.0

File: src/main/scala/com/amazon/deequ/analyzers/applicability/Applicability.scala (hunk 3 above)
Comment (radford1) [reply src/main/scala/com/amazon/deequ/analyzers/applicability/Applicability.scala:144]: This was a requirement to get spark 3.0 to work. It throws a long overflow exception in the checks.ApplicabilityTest.scala test when working with spark 3.0  I changed this code to operate in a valid timestamp range and it works after that. I tested with spark 2.x and it also passes the tests.

File: src/main/scala/com/amazon/deequ/constraints/AnalysisBasedConstraint.scala (hunk 4 above)
Comment (radford1) [reply src/main/scala/com/amazon/deequ/constraints/AnalysisBasedConstraint.scala:2]: Pulled this code over from PR #253 for scala 2.12 compatibility

File: src/main/scala/com/amazon/deequ/constraints/AnalysisBasedConstraint.scala (hunk 5 above)
Comment (radford1) [reply src/main/scala/com/amazon/deequ/constraints/AnalysisBasedConstraint.scala:43]: Pulled this code over from PR #253 for scala 2.12 compatibility

**Comment by radford1** [comment]
@sscdotopen I pulled some of this code over from PR #253 for scala 2.12 compatibility. The only change related to spark 3.0 was in src/main/scala/com/amazon/deequ/analyzers/applicability/Applicability.scala because of a long overflow error when running with spark 3.0. I can build and test this on different EMR versions. Any suggestion on the oldest EMR version to start with?

**Comment by radford1** [comment]
@sscdotopen I was able to package this on each of the following EMR versions
```
5.8.1 for spark 2.2 mvn package -P spark-2.2
5.14 for spark 2.3 mvn package -P spark-2.3
5.30 for spark 2.4 + scala 2.11 mvn package -P spark-2.4
6.0 for spark 2.4 + scala 2.12 mvn package -P spark-2.4,scala-2.12
6.1 for spark 3 + scala 2.12 mvn package -P spark-3.0,scala-2.12
```

**Comment by radford1** [comment]
I have finished testing on different EMR versions using the VerificationSuiteTest and VerificationResultTest with the following results:

```
EMR 5.8.1 PASS
EMR 5.14 PASS
EMR 5.30 Failed for the reuse existing results test case in the VerificationSuiteTest. I tested this against the master branch and it is also failing
EMR 6.0 PASS
EMR 6.1 Failed for the same reason as EMR 5.30
```

I have added the building of a test jar to continue to run these types of tests as well as a helper script called EMRSparkShellTest.scala in the test/resources folder with instructions on how to run it.

Please let me know if there is anything additionally that I can do to help push this over the finish line.

**Comment by tooptoop4** [comment]
gentle ping

**Comment by sscdotopen** [comment]
Thanks! Could you look at the CI failures and fix them?

**Comment by radford1** [comment]
Will take a look in the next few days

**Comment by radford1** [comment]
Fixed build errors which were all code style issues

**Comment by sscdotopen** [comment]
I would be ok with merging this, let me ping my colleagues to do the actual merge and release

**Comment by Teej42** [comment]
We are interested in this.  Please let us know what the merging/release version will be that will enable this support.  Thanks!

**Comment by rmalleman** [comment]
Any plans on releasing this anytime soon?
